### ISSUE TITLE
Interface like access to runtimes

### ISSUE THREAD

**Issue by mustermeiszer** [paritytech/substrate#13362]
As the discussion on the [forum](https://forum.polkadot.network/t/erc-20-like-standard-for-polkadot/1556) staled for some time I wanted to bring this the code side to actually move this forward.

The main idea is to support two things:
1. Submit extrinsics on an interface base rather than a pallet base &rarr; runtimes decide *what* the interface implementation actually executes
2. Query the chain state on an interface base via an runtime api similar to "view" methods on Ethereum

The requirements among other should be the following:
* Support para-chains, stand-alone chains and relaychains
* Support for XCM via `Transact`
* Integrate with the existing `construct_runtime!` macro
* Interfaces can be developed independently from external dependencies (i.e. this repo or any other team)

### Provide a new runtime API and two new traits
```rust
sp_api::decl_runtime_apis! {
	pub trait Interface {
		type View: View;

		fn view(view: Self::View) -> Result<Vec<u8>, InterfaceError> {
			view::view()
		}
	}
}

pub trait Core {
	type RuntimeOrigin;
}

pub trait View {
	fn view(&self) -> Result<Vec<u8>, InterfaceError>;
}
```

### Allow to define a new interface as a trait.
*Note: Interfaces maybe should NOT have associated types runtimes define, but rather agree on some primitive types to be used.*
```rust
#[frame_support::interface]
pub trait Pip20 {
   type AccountId;
   type Currency;
   type Balance;

   #[interface::view]
   #[interface::view_index(0)]
   fn free_balance(who: Self::AccountId, currency: Self::Currency) -> Balance;

   #[interface::call]
   #[interface::call_index(0)]
   fn transfer(
	origin: Self::RuntimeOrigin,
	recv: Self::AccountId,
	currency: Self::Currency,
	amount: Self::Balance,
    ) -> DispatchResult;
}
```

The procedural macro takes care of 
* Making `trait Core` a super trait of `trait Pip20`
* Creating an `enum View` -- implements among others `trait View`
* Creating an `enum Call` -- Implements among others `trait UnfilteredDispatchable`, `trait GetCallName` and `trait GetDispatchInfo`

The following shows an example expansion. Omitting trait implementation like `Encode` and `Decode` that would be needed.
```rust
pub mod __expanded {
	pub trait Pip20: Core {
		type AccountId;
		type Currency;
		type Balance;

		fn free_balance(who: Self::AccountId, currency: Self::Currency) -> Balance;

		fn transfer(
			origin: Self::RuntimeOrigin,
			recv: Self::AccountId,
			currency: Self::Currency,
			amount: Self::Balance,
		) -> DispatchResult;
	}

	pub enum View<Runtime> {
		FreeBalance { who: Runtime::AccountId, currency: Runtime::Currency },
	}

	impl<Runtime: Pip20 + Core> super::View for View<Runtime> {
		fn view(&self) -> Result<Vec<u8>, InterfaceError> {
			match self {
				View::FreeBalance { who, currency } =>
					Ok(<Runtime as Pip20>::free_balance(who, currency).encode()),
			}
		}
	}

	pub enum Call<Runtime> {
		Transfer {
			recv: Runtime::AccountId,
			currency: Runtime::Currency,
			amount: Runtime::Balance,
		},
	}

	impl<Runtime: Pip20 + Core> UnfilteredDispatchable for Call<Runtime> {
		type RuntimeOrigin = Runtime::RuntimeOrigin;

		fn dispatch_bypass_filter(
			self,
			origin: Self::RuntimeOrigin,
		) -> DispatchResultWithPostInfo {
			match self {
				Call::Transfer { recv, currency, amount } =>
					<Runtime as Pip20>::transfer(origin, recv, currency, amount),
			}
		}
	}

	impl<Runtime> GetDispatchInfo for Call<Runtime> {
		fn get_dispatch_info(&self) -> DispatchInfo {
			todo!()
		}
	}

	impl<Runtime> GetCallName for Call<Runtime> {
		fn get_call_names() -> &'static [&'static str] {
			todo!()
		}

		fn get_call_name(&self) -> &'static str {
			todo!()
		}
	}
}
```

### Runtime Adjustments
Runtimes that want to support interfaces must 
* Implement the needed trait
```rust
impl trait Core for Runtime {
   type RuntimeOrigin = RuntimeOrigin;
}

impl trait Pip20 for Runtime {
    type AccountId = AccountId;
    type Currency = Currency;
    type Balance = Balance; 
    
    fn free_balance(who: Self::AccountId, currency: Self::Currency) -> Balance {
        // Collect balances here for right currency
    }


    fn transfer(
	origin: Self::RuntimeOrigin,
	recv: Self::AccountId,
	currency: Self::Currency,
	amount: Self::Balance,
     ) -> DispatchResult {
         // Call pallets directly here to transfer right currency
     }
}
```
* Define a call interface
```rust
#[frame_support::call_interface]
pub enum CallInterface<Runtime> {
	#[call_index(20)]
	Pip20(pip20::Call<Runtime>),
}
```
* Define a view interface
```rust
#[frame_support::view_interface]
pub enum ViewInterface<Runtime> {
	#[call_index(20)]
	Pip20(pip20::View<Runtime>),
}
```
* Adjust their `construct_runtime!`
```rust
construct_runtime!(
	pub enum Runtime where
		Block = Block,
		NodeBlock = cfg_primitives::Block,
		UncheckedExtrinsic = UncheckedExtrinsic
                 Interface = CallInterface<Self>
	{
           ...
```
* Implement the runtime api
```rust
impl_runtime_apis! {
	impl Interface for Runtime {
		type View: ViewInterface<Self>;
	}
}
```

## Open Questions
* Weight handling

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/erc-20-like-standard-for-polkadot/1556/16

**Comment by bkchr** [comment]
For `view_function` I created today this issue: https://github.com/paritytech/substrate/issues/13358

**Comment by mustermeiszer** [comment]
@bkchr nice! But although the naming is the same, the view functions here serve a different purpose. They should rather abstract over pallet-specific access. If I understood @xlc correctly the view functions in #13358 serve to provide entrypoints into the wasm, while the once above serve to provide querying in companion to a specific interface. Similar to the once in the Ethereum world (e.g. ERC20), which are abstract over the explicit storage of the respective contracts. 

cc: @shawntabrizi to follow up on the stuff discussed in the forum.

**Comment by bkchr** [comment]
The view function are for querying the Pallet storage. Sounds exactly like what you describing here as well? Just that you want the querying functionality associated to an "Interface".

**Comment by mustermeiszer** [comment]
Yes, and that is a big difference. External parties (wallets, etc.) must now the specific pallet in use on every chain. In some cases, this will even be more complex, as one probably must know multiple locations to query in order to get the information they want. While when using the interface, the external party has the same entry point, the same scale-bytes they submit to every chain that uses that interface, while the runtime can decide how it is aggregating the right information based in the pallet it is using.

**Comment by mustermeiszer** [comment]
To be clear, I think the view functions that are drafted in your PR are really valuable and can be used to build an application around a chains wasm and more. But their are specific to a chain and the pallets it is using.

**Comment by bkchr** [comment]
Yeah, I mean you just want some standardized interface to fetch the same information on every chain implementing it, using a different implementation. This should be entirely doable with a runtime api? Calling via this interface isn't required to be supported. You could solve this by having functions that are returning you the calls or directly extrinsics. So you would have some `transfer` function that returns you the correct call/extrinsic to make a transfer.

**Comment by mustermeiszer** [comment]
> Yeah, I mean you just want some standardized interface to fetch the same information on every chain implementing it, using a different implementation. This should be entirely doable with a runtime api? Calling via this interface isn't required to be supported. You could solve this by having functions that are returning you the calls or directly extrinsics. So you would have some transfer function that returns you the correct call/extrinsic to make a transfer.

I really like the idea of having a runtime API that generates the payload to be signed and returns the extrinsics without the signature or so. This would make it much easier to generate extrinsics for external parties. 

But I think removing the interface from the call side has two downsides:
* Not XCM compatible
* Two-step process where it could be a one step process. I.e. fetch-calls & submit vs. submit

My main objective here is to make integrations with any chain easier. It would be really helpful if somebody who is building a wallet could give their two-cents here. Also, other parchain teams. We should be able to build something like ERC-20 or ERC-4626 in the ecosystem without relying on specific pallets or chain knowledge.

**Comment by bkchr** [comment]
> * Not XCM compatible

Why isn't it XCM compatible?

> * Two-step process where it could be a one step process. I.e. fetch-calls & submit vs. submit

You need to fetch some data anyway before, it is very unlikely that you can just submit an extrinsic. So, fetching some more data shouldn't be any problem, IMO.

**Comment by mustermeiszer** [comment]
> Why isn't it XCM compatible?
Just like the current `Transfer` instructions of the executor, the interface abstracts over the actual implementation. But if the interface is not part of the Call, then Chain A needs to know the exact behaviour of Chain B in order to call it via XCM.

> You need to fetch some data anyway before, it is very unlikely that you can just submit an extrinsic. So, fetching some more data shouldn't be any problem, IMO.

Agreed, thats a good point. 

## Next Steps
* Could something like that make it to the codebase?
* Who is good to talk to from the wallet/custodian side?

**Comment by bkchr** [comment]
> * Who is good to talk to from the wallet/custodian side?

CC @BenWhiteJam @joepetrowski @SBalaguer someone of you can help here?

**Comment by mustermeiszer** [comment]
Another think I realized: 
* "fine grained" access control would be necessary &rarr; similar to how Fireblocks allows to whitelist contracts that can be called

Maybe the macros can have something like the selector or so in Ethereum. E.g.:

```rust
#[frame_support::interface]
#[interface::discriminant(Self, Self::Currency)]
pub trait Pip20 {
      type AccountId = AccountId;
      type Currency = Currency;
      type Balance = Balance; 
   ...
}
```
and the discriminant `hash_of(Self, Currency)` would be then like a contract address wallets can whitelist again. 
Not sure, how this would tie together though...

**Comment by bkchr** [comment]
@mustermeiszer you brought up the idea, could you come up with some POC? Then we could see where we go from there?

**Comment by mustermeiszer** [comment]
@bkchr Happy to do that!

**Comment by kianenigma** [comment]
> @bkchr Happy to do that!

Looking forward!
